File: splaytree.m
:- module splaytree.

:- interface.

:- import_module list, pair.

:- type splaytree(K, V) ---> [] ; splayt(K, V, splaytree(K, V), splaytree(K, V)).

:- func new = splaytree(_, _).

:- func new(list(pair(K, V))) = splaytree(K, V) is semidet.

:- pred put(splaytree(K, V), K, V, splaytree(K, V)).
:- mode put(in, in, in, out) is semidet.

:- pred get(splaytree(K, V), K, V, splaytree(K, V)).
:- mode get(in, in, out, out) is semidet.

:- pred put_list(splaytree(K, V), list(pair(K, V)), splaytree(K, V)).
:- mode put_list(in, in, out) is semidet.

:- implementation.

new = [].

new(L) = T :-
        T1 = new,
        put_list(T1, L, T).

:- pred put1(comparison_result, K, V, splaytree(K, V), splaytree(K, V), K, splaytree(K, V), splaytree(K, V)).
:- mode put1(in, in, in, in, in, in, out, out) is semidet.

put1((<), KO, VO, LO, RO, K, L1, R) :-
        put1(LO, K, L1, R1),
        R = splayt(KO, VO, R1, RO).

put1((>), KO, VO, LO, RO, K, L, R1) :-
        put1(RO, K, L1, R1),
        L = splayt(KO, VO, LO, L1).

:- pred put1(splaytree(K, V), K, splaytree(K, V), splaytree(K, V)).
:- mode put1(in, in, out, out) is semidet.

put1([], _, [], []).
put1(splayt(KO, VO, LO, RO), K, L, R) :-
        compare(O, K, KO),
        put1(O, KO, VO, LO, RO, K, L, R).

put(T, K, V, Out) :-
        put1(T, K, L, R),
        Out = splayt(K, V, L, R).

put_list(T, [], T).
put_list(T, [K-V|L], T1) :-
        put(T, K, V, T2),
        put_list(T2, L, T1).


:- pred get1(comparison_result, K, V, splaytree(K, V), splaytree(K, V), K, V, splaytree(K, V), splaytree(K, V)).
:- mode get1(in, in, in, in, in, in, out, out, out).

get1((=), K, V, L, R, K, V, L, R).
get1((<), KO, VO, splayt(KLO, VLO, LLO, RLO), RO, K, V, L1, R) :-
        get1(KLO, VLO, LLO, RLO, K, V, L1, R1),
        R = splayt(KO, VO, R1, RO).
get1((>), KO, VO, LO, splayt(KRO, VRO, LRO, RRO), K, V, L, R1) :-
        get1(KRO, VRO, LRO, RRO, K, V, L1, R1),
        L = splayt(KO, VO, LO, L1).

:- pred get1(K, V, splaytree(K, V), splaytree(K, V), K, V, splaytree(K, V), splaytree(K, V)).
:- mode get1(in, in, in, in, in, out, out, out).

get1(KO, VO, LO, RO, K, V, L, R) :-
        compare(O, K, KO),
        get1(O, KO, VO, LO, RO, K, V, L, R).

get(splayt(KO, VO, LO, RO), K, V, Out) :-
        get1(KO, VO, LO, RO, K, V, L, R),
        Out = splayt(K, V, L, R).

File: splaytree.pl
% -*- Mode: Prolog -*-

:- module(splaytree,
          [spt_new/1,
           spt_put/4,
           spt_get/4,
           list_to_spt/2,
           spt_maplist/4,
           spt_dump/1,

           sptw_new/1,
           sptw_put/3,
           sptw_get/3,
           list_to_sptw/2,
           sptw_maplist/3,
           sptw_dump/1
           ]).

spt_new(spt).

spt_put(T, K, V, spt(K, V, L, R)) :-
        put(T, K, L, R).

put(spt, _, spt, spt).

put(spt(KO, VO, LO, RO), K, L, R) :-
        compare(O, K, KO),
        put(O, KO, VO, LO, RO, K, L, R).

put(<, KO, VO, LO, RO, K, L1, spt(KO, VO, R1, RO)) :-
        put(LO, K, L1, R1).

put(>, KO, VO, LO, RO, K, spt(KO, VO, LO, L1), R1) :-
        put(RO, K, L1, R1).

spt_get(spt(KO, VO, LO, RO), K, V, spt(K, V, L, R)) :-
        get(KO, VO, LO, RO, K, V, L, R).

get(KO, VO, LO, RO, K, V, L, R) :-
        compare(O, K, KO),
        get(O, KO, VO, LO, RO, K, V, L, R).

get(=, K, V, L, R, K, V, L, R).

get(<, KO, VO, spt(KLO, VLO, LLO, RLO), RO, K, V, L1, spt(KO, VO, R1, RO)) :-
        get(KLO, VLO, LLO, RLO, K, V, L1, R1).

get(>, KO, VO, LO, spt(KRO, VRO, LRO, RRO), K, V, spt(KO, VO, LO, L1), R1) :-
        get(KRO, VRO, LRO, RRO, K, V, L1, R1).


list_to_spt(L, T) :-
        spt_new(T1),
        append_list(L, T1, T).

append_list([], T, T).
append_list([K-V|L], T, T1) :-
        spt_put(T, K, V, T2),
        append_list(L, T2, T1).

spt_maplist(T, L, L1, T1) :-
        maplist(L, T, L1, T1).

maplist([], T, [], T).
maplist([K|L], T, [K-V|L1], T1) :-
       spt_get(T, K, V, T2),
       maplist(L, T2, L1, T1).

sptw_new(sptw(T)) :-
        spt_new(T).

sptw_put(W, K, V) :-
        W = sptw(T),
        spt_put(T, K, V, T1),
        setarg(1, W, T1).

sptw_get(W, K, V) :-
        W = sptw(T),
        spt_get(T, K, V, T1),
        setarg(1, W, T1).

list_to_sptw(L, sptw(T)) :-
        list_to_spt(L, T).

sptw_maplist(W, L, M) :-
        W = sptw(T),
        spt_maplist(T, L, M, T1),
        setarg(1, W, T1).

spt_dump(T) :-
	spt_dump(T, '').
spt_dump(spt, S) :-
	format('~pspt~n', [S]).
spt_dump(spt(K, V, L, R), S) :-
	format('~pspt ~p=~p~n', [S, K, V]),
	atom_concat(S, '   |', SL),
	spt_dump(L, SL),
	atom_concat(S, '    ', SR),
	spt_dump(R, SR).

sptw_dump(sptw(T)) :-
        spt_dump(T).